hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/commit/HoodieMergeHelper.java

```diff
@@ -202,7 +202,9 @@ private Option<Function<HoodieRecord, HoodieRecord>> composeSchemaEvolutionTrans
       Schema newWriterSchema = AvroInternalSchemaConverter.convert(mergedSchema, writerSchema.getFullName());
       Schema writeSchemaFromFile = AvroInternalSchemaConverter.convert(writeInternalSchema, newWriterSchema.getFullName());
       boolean needToReWriteRecord = sameCols.size() != colNamesFromWriteSchema.size()
-          || SchemaCompatibility.checkReaderWriterCompatibility(newWriterSchema, writeSchemaFromFile).getType() == org.apache.avro.SchemaCompatibility.SchemaCompatibilityType.COMPATIBLE;
+          && SchemaCompatibility.checkReaderWriterCompatibility(newWriterSchema, writeSchemaFromFile).getType()
+          == org.apache.avro.SchemaCompatibility.SchemaCompatibilityType.COMPATIBLE;
+
```

Comment:
@danny0405  @ThinkerLei 
i think again.
SameCols. size() == colNamesFromWriteScheme. size() only happen in following scence
The table has new columns, while the old columns have not been changed(rename, type change).
eg:
```
write schema: a string, b int, c long
read schema: a string, b int, c long, d int
```
In this case
SameCols. size() == colNamesFromWriteScheme. size().
and, writeSchema is equivalent to a pruned readschema.

However, some versions of AVRO, such as AVRO 1.8. x , may report errors when using pruned schemas to read AVRO files. (avro 1.10x has no such problem)

Therefore, even if sameCols. size() == colNamesFromWriteScheme. size(), we still need to check the compatibility of the read-write schema. If it is compatible, we can directly use this writeSchema to read avo data.


Therefore, maybe we can use following logic to avoid unnecessary rewrite.
```
boolean needToReWriteRecord =  sameCols.size() != colNamesFromWriteSchema.size() || 
!SchemaCompatibility.checkReaderWriterCompatibility(newWriterSchema, writeSchemaFromFile).getType() == org.apache.avro.SchemaCompatibility.SchemaCompatibilityType.COMPATIBLE
```

Reply:
Good point for optimization, we introduce some changes like the dynamic read schema based on write schema in release 1.x as for the `HoodieFileGroupReader`, but I'm not sure whether it is applied automically for all the read paths, cc @yihua for confirming this.

And anyway, I think we should have such optimization in 0.x branch and master for the legacy `HoodieMergedLogRecordReader` which will still be benefic to engines line Flink and Hive.

@xiarixiaoyao do you have intreast to contribute this?

Reply:
> @danny0405 @ThinkerLei i think again. SameCols. size() == colNamesFromWriteScheme. size() only happen in following scence The table has new columns, while the old columns have not been changed(rename, type change). eg:
> 
> ```
> write schema: a string, b int, c long
> read schema: a string, b int, c long, d int
> ```
> 
> In this case SameCols. size() == colNamesFromWriteScheme. size(). and, writeSchema is equivalent to a pruned readschema.
> 
> However, some versions of AVRO, such as AVRO 1.8. x , may report errors when using pruned schemas to read AVRO files. (avro 1.10x has no such problem)
> 
> Therefore, even if sameCols. size() == colNamesFromWriteScheme. size(), we still need to check the compatibility of the read-write schema. If it is compatible, we can directly use this writeSchema to read avo data.
> 
> maybe we can use following logic to avoid unnecessary rewrite.
> 
> ```
> boolean needToReWriteRecord =  sameCols.size() != colNamesFromWriteSchema.size() || 
> !SchemaCompatibility.checkReaderWriterCompatibility(newWriterSchema, writeSchemaFromFile).getType() == org.apache.avro.SchemaCompatibility.SchemaCompatibilityType.COMPATIBLE
> ```

This was the logic I initially fixed. Do I still need to make changes based on this PR? cc @danny0405 @xiarixiaoyao

Reply:
Yeah, let's change it back, we better have some test cases.

Reply:
> Good point for optimization, we introduce some changes like the dynamic read schema based on write schema in release 1.x as for the `HoodieFileGroupReader`, but I'm not sure whether it is applied automically for all the read paths, cc @yihua for confirming this.
> 
> And anyway, I think we should have such optimization in 0.x branch and master for the legacy `HoodieMergedLogRecordReader` which will still be benefic to engines line Flink and Hive.
> 
> @xiarixiaoyao do you have intreast to contribute this?

will try

Reply:
@xiarixiaoyao This info is valuable.  Basically using pruned schema to read Avro records is supported on Avro 1.10 and above, not on lower versions.  I see that Spark 3.2 and above and all Flink versions use Avro 1.10 and above.  So for these integrations and others that rely on Avro 1.10 and above, we should use pruned schema to read log records to improve performance.  I'll check the new file group reader.